sentry-core/src/main/java/io/sentry/core/cache/CacheStrategy.java

```diff
@@ -91,4 +118,192 @@ protected void rotateCacheIfNeeded(final @NotNull File[] files) {
+    final SentryEnvelope currentEnvelope = readEnvelope(currentFile);
+
+    if (!isValidEnvelope(currentEnvelope)) {
+      return;
+    }
+
+    final Session currentSession = getFirstSession(currentEnvelope);
+
+    if (!isValidSession(currentSession)) {
+      return;
+    }
+
+    // nothing to do if its not true
+    final Boolean currentSessionInit = currentSession.getInit();
+    if (currentSessionInit == null || !currentSessionInit) {
+      return;
+    }
+
+    // we need to move the init flag
+    for (final File notDeletedFile : notDeletedFiles) {
+      final SentryEnvelope envelope = readEnvelope(notDeletedFile);
+
+      if (!isValidEnvelope(envelope)) {
+        continue;
+      }
+
+      SentryEnvelopeItem newSessionItem = null;
+      final Iterator<SentryEnvelopeItem> itemsIterator = envelope.getItems().iterator();
+
+      while (itemsIterator.hasNext()) {
+        final SentryEnvelopeItem envelopeItem = itemsIterator.next();
+
+        if (!isSessionType(envelopeItem)) {
+          continue;
+        }
+
+        final Session session = readSession(envelopeItem);
+
+        if (!isValidSession(session)) {
+          continue;
+        }
+
+        // if its already true, nothing to do
+        final Boolean init = session.getInit();
+        if (init != null && init) {
+          return;
+        }
+
+        if (currentSession.getSessionId().equals(session.getSessionId())) {
+          session.setInitAsTrue();
+          try {
+            newSessionItem = SentryEnvelopeItem.fromSession(serializer, session);
+            // remove item from envelope items so we can replace with the new one that has the
+            // init flag true
+            itemsIterator.remove();
+          } catch (IOException e) {
+            options
+                .getLogger()
+                .log(
+                    ERROR,
+                    e,
+                    "Failed to create new envelope item for the currentSession %s",
+                    currentSession.getSessionId());
+          }
+
+          break;
+        }
+      }
+
+      if (newSessionItem != null) {
+        final SentryEnvelope newEnvelope = buildNewEnvelope(envelope, newSessionItem);
+
+        long notDeletedFileTimestamp = notDeletedFile.lastModified();
+        if (!notDeletedFile.delete()) {
+          options
+              .getLogger()
+              .log(
+                  SentryLevel.WARNING,
+                  "File can't be deleted: %s",
+                  notDeletedFile.getAbsolutePath());
+        }
+
+        saveNewEnvelope(newEnvelope, notDeletedFile, notDeletedFileTimestamp);
+        break;
+      } else {
+        // TODO: if theres no more files with that session id, should we still delete it?
+        // that means the next session update will be init=false but the one
+        // with init=true will be deleted.
+        // either we don't delete it or we generate a new envelope only with that session.
```

Comment:
@bruno-garcia I added this comment just to discuss it here, but this is an acceptable thing as its a corner case, right?

Reply:
yeah lets creeate something on asana to track but it's fine to just drop it.

I was going to comment to look into the open session but that's racey and anyway too much complexity to cover this. Happy with just removing it in this case

Reply:
actually this situation is described in the DACI, right, do we really need to create a ticket out of it? I mean, its something that we agreed on